### PR TITLE
Add DaxpyIncTo

### PR DESCRIPTION
**Pull request by vladimir-ch**
Replacement of #13

Shortly I will submit PRs to packages that are affected by this PR, and will not merge until they are all LGTMed.

### CODE REVIEW

File: asm/daxpy_amd64.s [hunk 1]
```diff
@@ -87,7 +87,7 @@ TEXT ·DaxpyInc(SB), NOSPLIT, $0
 	MOVLPD alpha+0(FP), X7
 	MOVQ   x+8(FP), R8
 	MOVQ   y+32(FP), R9
-	MOVQ   n+56(FP), CX
+	MOVQ   n+56(FP), DX
```
Comment (vladimir-ch) [inline asm/daxpy_amd64.s:91]: I switch to DX here to be consistent with `DaxpyIncTo`.

File: asm/daxpy_test.go [hunk 2]
```diff
@@ -261,33 +202,154 @@ func TestDaxpyInc(t *testing.T) {
...
+	x := randomSlice(n, 1)
+	y := randomSlice(n, 1)
+	dst := randomSlice(n, 1)
 	a := rand.Float64()
 	b.ResetTimer()
 	for i := 0; i < b.N; i++ {
-		DaxpyUnitary(a, x, y, z)
-		gs = z
+		DaxpyUnitaryTo(dst, a, x, y)
+		gs = dst
```
Comment (kortschak) [inline asm/daxpy_test.go]: You can do this just once.

File: asm/daxpy_amd64.s [hunk 3]
```diff
@@ -41,22 +41,22 @@
...
+	MOVQ   x+32(FP), R8
+	MOVQ   x_len+40(FP), DI // n = len(x)
+	MOVQ   y+56(FP), R9
 
 	MOVQ $0, SI // i = 0
 	SUBQ $2, DI // n -= 2
 	JL   V1     // if n < 0 goto V1
 
 U1:  // n >= 0
-	// y[i] += alpha * x[i] unrolled 2x.
```
Comment (kortschak) [inline asm/daxpy_amd64.s:59]: This looks like it was previously wrong. True?

**Comment by kortschak** [comment]
It's worth adding the go generate hurdle to the tests.

File: asm/daxpy_amd64.s [hunk 4]
```diff
@@ -136,3 +136,67 @@ V2:
...
+	ADDQ R12, BX // nextY += incY
+	ADDQ R13, CX // nextDst += incDst
+	SHLQ $1, R11 // incX *= 2
+	SHLQ $1, R12 // incY *= 2
+	SHLQ $1, R13 // incDst *= 2
+
+	SUBQ $2, DX // n -= 2
+	JL   V3     // if n < 0 goto V2
+
+U3:  // n >= 0
```
Comment (kortschak) [inline asm/daxpy_amd64.s:175]: Was asmfmt run?

File: asm/daxpy_amd64.s [hunk 5]
```diff
@@ -136,3 +136,67 @@ V2:
...
+	MOVLPD alpha+40(FP), X7
+	MOVQ   x+48(FP), R8
+	MOVQ   y+72(FP), R9
+	MOVQ   n+96(FP), DX
+	MOVQ   incX+104(FP), R11
+	MOVQ   incY+112(FP), R12
+	MOVQ   ix+120(FP), SI
+	MOVQ   iy+128(FP), DI
+
+	MOVQ SI, AX  // nextX = ix
```
Comment (kortschak) [inline asm/daxpy_amd64.s:162]: I take it that the code here was copy/past from DaxpyInc except for the preamble? Can we have either a comment to this effect or something that generates both?

File: asm/daxpy_test.go [hunk 6]
```diff
@@ -348,7 +453,19 @@ func daxpyIncBenchmark(b *testing.B, n, inc int) {
 	a := rand.Float64()
 	b.ResetTimer()
 	for i := 0; i < b.N; i++ {
-		DaxpyInc(a, x, y, uintptr(n), uintptr(inc), uintptr(inc), uintptr(ini), uintptr(ini))
+		DaxpyIncTo(dst, uintptr(inc), uintptr(ini), a, x, y,
+			uintptr(n), uintptr(inc), uintptr(inc), uintptr(ini), uintptr(ini))
 		gs = y
```
Comment (kortschak) [inline asm/daxpy_test.go]: Do this once.

File: asm/daxpy_amd64.s (hunk 3 above)
Comment (vladimir-ch) [inline asm/daxpy_amd64.s:59]: Yes, the comment was wrong.

File: asm/daxpy_amd64.s (hunk 4 above)
Comment (vladimir-ch) [inline asm/daxpy_amd64.s:175]: Yes, asmfmt gives this formatting but it is true that having one space less would be nicer (this particular comment is not very useful, I left it just to keep the changes at minimum). I will look into it.

Comment (kortschak) [inline asm/daxpy_amd64.s:175]: ... or one space more.

File: asm/daxpy_amd64.s (hunk 5 above)
Comment (vladimir-ch) [inline asm/daxpy_amd64.s:162]: Yes, it was copy/paste plus modified preamble and loop body (updates to BP and CX). I will write a comment with a TODO for the moment and consider generating this after this group of PRs is merged.

Comment (kortschak) [inline asm/daxpy_amd64.s:162]: SGTM

File: asm/daxpy_amd64.s (hunk 4 above)
Comment (vladimir-ch) [inline asm/daxpy_amd64.s:175]: How wide are your tabs? :-)

Comment (kortschak) [inline asm/daxpy_amd64.s:175]: I was reading on github.

File: asm/daxpy_test.go (hunk 2 above)
Comment (vladimir-ch) [inline asm/daxpy_test.go]: Done

File: asm/daxpy_test.go (hunk 6 above)
Comment (vladimir-ch) [inline asm/daxpy_test.go]: Done.

**Comment by vladimir-ch** [comment]
PTAL (especially at the travis thing)

File: .travis.yml [hunk 7]
```diff
@@ -16,9 +17,11 @@ before_install:
...
  - go get -d -t -v ./...
- - go build -v ./...
- - go test -v -a ./...
- - go test -v -a -tags noasm ./...
- - go test -v -a -tags appengine ./...
+ - go build -v -x ./...
+ - go test -v -x -a ./...
+ - go test -v -x -a -tags noasm ./...
+ - go test -v -x -a -tags appengine ./...
  - diff <(gofmt -d .) <("")
```
Comment (kortschak) [inline .travis.yml:25]: Is it worth adding the same check for asmfmt?

File: .travis/check-generate.sh [hunk 8]
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+
+go generate github.com/gonum/matrix
```
Comment (kortschak) [inline .travis/check-generate.sh]: This should be `go generate github.com/gonum/internal/asm`.

**Comment by kortschak** [comment]
LGTM after go generate invocation fixed.

File: .travis/check-generate.sh (hunk 8 above)
Comment (vladimir-ch) [inline .travis/check-generate.sh]: Ooops, sorry, just blindly copying things around.

File: .travis.yml (hunk 7 above)
Comment (vladimir-ch) [inline .travis.yml:25]: Good idea but perhaps we should wait until it stabilizes a bit? I will add a TODO.

Comment (kortschak) [inline .travis.yml:25]: No worries.

**Pull request merged**